docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,448 @@
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instances are instantiated and initialized. 
+
+# System architecture and high-level design
+
+## Design basics
+
+SecureStore is a storage class, derived from KVStore. It adds security features to the underlying key value store.
+ 
+As such, it offers all KVStore APIs, with additional security options (which can be selected using the creation flags at set). These include:  
+- Encryption: Data is encrypted using the AES-CTR encryption method, where the IV consists of a randomly generated 8-byte nonce and an 8-byte counter. Key is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is the key trimmed to 32 bytes, with "ENC" as prefix.  
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data. The key used for generating the CMAC is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is the key trimmed to 32 bytes, with "AUTH" as prefix.  
+- Rollback protection: (Requires authentication) CMAC is stored in a designated rollback protected storage (also of KVStore type) and compared to when reading the data under the same key.
+- Write once: Key can only be stored once and can't be removed. 
+
+
+![SecureStore Layers](./SecureStore_layers.jpg)
+
+
+### Data layout
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
+
+![SecureStore Record](./SecureStore_record.jpg)
+
+Fields are:
+- Metadata size: Size of metadata header
+- Revision: SecureStore revision (currently 1)
+- Data size: Size of user data
+- Flags: User flags
+- Nonce: Nonce field used in IV
+- Pad: Pad data to a multiple of 16 bytes (due to encryption)
+- CMAC: CMAC calculated on key, metadata & data  
+
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP keys are the same as the SecureStore keys. 
```

Comment:
Maybe we need to make a distinction between cryptographic keys and KVStore keys. I had to read this section a couple of times

Reply:
Thanks. Will rephrase to make this clearer.